crates/ruma-identifiers/src/server_name.rs

```diff
@@ -139,6 +140,20 @@ impl<'de> serde::Deserialize<'de> for Box<ServerName> {
     }
 }
 
+impl From<&ServerName> for Rc<ServerName> {
+    fn from(s: &ServerName) -> Rc<ServerName> {
+        let rc = Rc::<str>::from(s.as_str());
+        unsafe { Rc::from_raw(Rc::into_raw(rc) as *const ServerName) }
+    }
+}
+impl From<&ServerName> for Arc<ServerName> {
+    fn from(s: &ServerName) -> Arc<ServerName> {
+        let arc = Arc::<str>::from(s.as_str());
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as *const ServerName) }
+    }
+}
```

Comment:
As mentioned on #615, there should be no `From` implementations, but `TryFrom` (and maybe `FromStr`). These should mirror the existing implementations on lines 102 through 116 in signature (with `Box` replaced by `Rc` / `Arc`), with the implementation given here but an up-front `validate(server_name.as_ref())?;` like in the free `fn try_from`.